crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1000,17 +973,15 @@ impl<'db> Type<'db> {
 
             (Type::Intersection(intersection), other)
             | (other, Type::Intersection(intersection)) => {
-                if intersection
+                intersection
                     .positive(db)
                     .iter()
                     .any(|p| p.is_disjoint_from(db, other))
-                {
-                    true
-                } else {
-                    // TODO we can do better here. For example:
-                    // X & ~Literal[1] is disjoint from Literal[1]
-                    false
-                }
+                    // A & B & Not[C] is disjoint from C
+                    || intersection
+                        .negative(db)
+                        .iter()
+                        .any(|&neg_ty| other.is_assignable_to(db, neg_ty))
```

Comment:
I think using `is_assignable_to` here is not right, but using `is_subtype_of` might be right? (See my comment above that `Any` and `Not[Any]` are not disjoint.)